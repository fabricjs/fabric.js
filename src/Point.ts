import type { TMat2D, TRadian } from './typedefs';
import { cos } from './util/misc/cos';
import { sin } from './util/misc/sin';

export interface XY {
  x: number;
  y: number;
}

/**
 * Adaptation of work of Kevin Lindsey(kevin@kevlindev.com)
 */
export class Point implements XY {
  declare x: number;

  declare y: number;

  constructor();
  constructor(x: number, y: number);
  constructor(point?: XY);
  constructor(arg0: number | XY = 0, y = 0) {
    if (typeof arg0 === 'object') {
      this.x = arg0.x;
      this.y = arg0.y;
    } else {
      this.x = arg0;
      this.y = y;
    }
  }

  /**
   * Adds another point to this one and returns another one
   * @param {XY} that
   * @return {Point} new Point instance with added values
   */
  add(that: XY): Point {
    return new Point(this.x + that.x, this.y + that.y);
  }

  /**
   * Adds another point to this one
   * @param {XY} that
   * @return {Point} thisArg
   * @chainable
   * @deprecated
   */
  addEquals(that: XY): Point {
    this.x += that.x;
    this.y += that.y;
    return this;
  }

  /**
   * Adds value to this point and returns a new one
   * @param {Number} scalar
   * @return {Point} new Point with added value
   */
  scalarAdd(scalar: number): Point {
    return new Point(this.x + scalar, this.y + scalar);
  }

  /**
   * Adds value to this point
   * @param {Number} scalar
   * @return {Point} thisArg
   * @chainable
   * @deprecated
   */
  scalarAddEquals(scalar: number): Point {
    this.x += scalar;
    this.y += scalar;
    return this;
  }

  /**
   * Subtracts another point from this point and returns a new one
   * @param {XY} that
   * @return {Point} new Point object with subtracted values
   */
  subtract(that: XY): Point {
    return new Point(this.x - that.x, this.y - that.y);
  }

  /**
   * Subtracts another point from this point
   * @param {XY} that
   * @return {Point} thisArg
   * @chainable
   * @deprecated
   */
  subtractEquals(that: XY): Point {
    this.x -= that.x;
    this.y -= that.y;
    return this;
  }

  /**
   * Subtracts value from this point and returns a new one
   * @param {Number} scalar
   * @return {Point}
   */
  scalarSubtract(scalar: number): Point {
    return new Point(this.x - scalar, this.y - scalar);
  }

  /**
   * Subtracts value from this point
   * @param {Number} scalar
   * @return {Point} thisArg
   * @chainable
   * @deprecated
   */
  scalarSubtractEquals(scalar: number): Point {
    this.x -= scalar;
    this.y -= scalar;
    return this;
  }

  /**
   * Multiplies this point by another value and returns a new one
   * @param {XY} that
   * @return {Point}
   */
  multiply(that: XY): Point {
    return new Point(this.x * that.x, this.y * that.y);
  }

  /**
   * Multiplies this point by a value and returns a new one
   * @param {Number} scalar
   * @return {Point}
   */
  scalarMultiply(scalar: number): Point {
    return new Point(this.x * scalar, this.y * scalar);
  }

  /**
   * Multiplies this point by a value
   * @param {Number} scalar
   * @return {Point} thisArg
   * @chainable
   * @deprecated
   */
  scalarMultiplyEquals(scalar: number): Point {
    this.x *= scalar;
    this.y *= scalar;
    return this;
  }

  /**
   * Divides this point by another and returns a new one
   * @param {XY} that
   * @return {Point}
   */
  divide(that: XY): Point {
    return new Point(this.x / that.x, this.y / that.y);
  }

  /**
   * Divides this point by a value and returns a new one
   * @param {Number} scalar
   * @return {Point}
   */
  scalarDivide(scalar: number): Point {
    return new Point(this.x / scalar, this.y / scalar);
  }

  /**
   * Divides this point by a value
   * @param {Number} scalar
   * @return {Point} thisArg
   * @chainable
   * @deprecated
   */
  scalarDivideEquals(scalar: number): Point {
    this.x /= scalar;
    this.y /= scalar;
    return this;
  }

  /**
   * Returns true if this point is equal to another one
   * @param {XY} that
   * @return {Boolean}
   */
  eq(that: XY): boolean {
    return this.x === that.x && this.y === that.y;
  }

  /**
   * Returns true if this point is less than another one
   * @param {XY} that
   * @return {Boolean}
   */
  lt(that: XY): boolean {
    return this.x < that.x && this.y < that.y;
  }

  /**
   * Returns true if this point is less than or equal to another one
   * @param {XY} that
   * @return {Boolean}
   */
  lte(that: XY): boolean {
    return this.x <= that.x && this.y <= that.y;
  }

  /**

   * Returns true if this point is greater another one
   * @param {XY} that
   * @return {Boolean}
   */
  gt(that: XY): boolean {
    return this.x > that.x && this.y > that.y;
  }

  /**
   * Returns true if this point is greater than or equal to another one
   * @param {XY} that
   * @return {Boolean}
   */
  gte(that: XY): boolean {
    return this.x >= that.x && this.y >= that.y;
  }

  /**
   * Returns new point which is the result of linear interpolation with this one and another one
   * @param {XY} that
   * @param {Number} t , position of interpolation, between 0 and 1 default 0.5
   * @return {Point}
   */
  lerp(that: XY, t = 0.5): Point {
    t = Math.max(Math.min(1, t), 0);
    return new Point(
      this.x + (that.x - this.x) * t,
      this.y + (that.y - this.y) * t
    );
  }

  /**
   * Returns distance from this point and another one
   * @param {XY} that
   * @return {Number}
   */
  distanceFrom(that: XY): number {
    const dx = this.x - that.x,
      dy = this.y - that.y;
    return Math.sqrt(dx * dx + dy * dy);
  }

  /**
   * Returns the point between this point and another one
   * @param {XY} that
   * @return {Point}
   */
  midPointFrom(that: XY): Point {
    return this.lerp(that);
  }

  /**
   * Returns a new point which is the min of this and another one
   * @param {XY} that
   * @return {Point}
   */
  min(that: XY): Point {
    return new Point(Math.min(this.x, that.x), Math.min(this.y, that.y));
  }

  /**
   * Returns a new point which is the max of this and another one
   * @param {XY} that
   * @return {Point}
   */
  max(that: XY): Point {
    return new Point(Math.max(this.x, that.x), Math.max(this.y, that.y));
  }

  /**
   * Returns string representation of this point
   * @return {String}
   */
  toString(): string {
    return `${this.x},${this.y}`;
  }

  /**
   * Sets x/y of this point
   * @param {Number} x
   * @param {Number} y
   * @chainable
   */
  setXY(x: number, y: number) {
    this.x = x;
    this.y = y;
    return this;
  }

  /**
   * Sets x of this point
   * @param {Number} x
   * @chainable
   */
  setX(x: number) {
    this.x = x;
    return this;
  }

  /**
   * Sets y of this point
   * @param {Number} y
   * @chainable
   */
  setY(y: number) {
    this.y = y;
    return this;
  }

  /**
   * Sets x/y of this point from another point
   * @param {XY} that
   * @chainable
   */
  setFromPoint(that: XY) {
    this.x = that.x;
    this.y = that.y;
    return this;
  }

  /**
   * Swaps x/y of this point and another point
   * @param {XY} that
   */
  swap(that: XY) {
    const x = this.x,
      y = this.y;
    this.x = that.x;
    this.y = that.y;
    that.x = x;
    that.y = y;
  }

  /**
   * return a cloned instance of the point
   * @return {Point}
   */
  clone(): Point {
    return new Point(this.x, this.y);
  }

  /**
   * Rotates `point` around `origin` with `radians`
   * @static
   * @memberOf fabric.util
   * @param {XY} origin The origin of the rotation
   * @param {TRadian} radians The radians of the angle for the rotation
   * @return {Point} The new rotated point
   */
  rotate(radians: TRadian, origin: XY = ZERO): Point {
    // TODO benchmark and verify the add and subtract how much cost
    // and then in case early return if no origin is passed
    const sinus = sin(radians),
      cosinus = cos(radians);
    const p = this.subtract(origin);
    const rotated = new Point(
      p.x * cosinus - p.y * sinus,
      p.x * sinus + p.y * cosinus
    );
    return rotated.add(origin);
  }

  /**
   * Apply transform t to point p
   * @static
   * @memberOf fabric.util
   * @param  {TMat2D} t The transform
   * @param  {Boolean} [ignoreOffset] Indicates that the offset should not be applied
   * @return {Point} The transformed point
   */
  transform(t: TMat2D, ignoreOffset = false): Point {
    return new Point(
      t[0] * this.x + t[2] * this.y + (ignoreOffset ? 0 : t[4]),
      t[1] * this.x + t[3] * this.y + (ignoreOffset ? 0 : t[5])
    );
  }
}

export const ZERO = new Point(0, 0);
