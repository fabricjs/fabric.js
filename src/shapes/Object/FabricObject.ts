import { ObjectEvents } from '../../EventTypeDefs';
import { FabricObjectSVGExportMixin } from './FabricObjectSVGExportMixin';
import { InteractiveFabricObject } from './InteractiveObject';
import { applyMixins } from '../../util/applyMixins';
import { StatefulMixin } from '../../mixins/stateful.mixin';

// TODO somehow we have to make a tree-shakeable import

// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unused-vars
export interface FabricObject<EventSpec extends ObjectEvents = ObjectEvents>
  extends FabricObjectSVGExportMixin {}

export class FabricObject<
  EventSpec extends ObjectEvents = ObjectEvents
> extends InteractiveFabricObject<EventSpec> {}

applyMixins(FabricObject, [FabricObjectSVGExportMixin, StatefulMixin]);

export { cacheProperties, stateProperties } from './defaultValues';
