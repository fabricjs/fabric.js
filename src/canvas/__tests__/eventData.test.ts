/* eslint-disable no-restricted-globals */
import '../../../jest.extend';
import type { TPointerEvent } from '../../EventTypeDefs';
import { Point } from '../../Point';
import { Group } from '../../shapes/Group';
import { IText } from '../../shapes/IText/IText';
import { FabricObject } from '../../shapes/Object/FabricObject';
import type { TMat2D } from '../../typedefs';
import { Canvas } from '../Canvas';

const genericVpt = [2.3, 0, 0, 2.3, 120, 80] as TMat2D;

describe('Canvas event data', () => {
  let canvas: Canvas;
  let spy: jest.SpyInstance;

  const snapshotOptions = {
    cloneDeepWith: (value: any) => {
      if (value instanceof Point) {
        return new Point(Math.round(value.x), Math.round(value.y));
      }
    },
  };

  beforeEach(() => {
    canvas = new Canvas();
    spy = jest.spyOn(canvas, 'fire');
  });

  afterEach(() => {
    return canvas.dispose();
  });

  test.each([
    'mousedown',
    'mouseenter',
    'mouseleave',
    'mousemove',
    'mouseout',
    'mouseover',
    'dblclick',
    'wheel',
    'contextmenu',
  ] as (keyof WindowEventMap)[])(
    'HTML event "%s" should fire a corresponding canvas event',
    (type) => {
      canvas.setViewportTransform(genericVpt);
      canvas
        .getSelectionElement()
        .dispatchEvent(new MouseEvent(type, { clientX: 50, clientY: 50 }));
      expect(spy.mock.calls).toMatchSnapshot(snapshotOptions);
    }
  );

  // must call mousedown for mouseup to be listened to
  test('HTML event "mouseup" should fire a corresponding canvas event', () => {
    canvas.setViewportTransform(genericVpt);
    canvas
      .getSelectionElement()
      .dispatchEvent(new MouseEvent('mousedown', { clientX: 50, clientY: 50 }));
    spy.mockReset();
    document.dispatchEvent(
      new MouseEvent('mouseup', { clientX: 50, clientY: 50 })
    );
    expect(spy.mock.calls).toMatchSnapshot(snapshotOptions);
  });

  test.each([
    'drag',
    'dragend',
    'dragenter',
    'dragleave',
    'dragover',
    'drop',
  ] as (keyof WindowEventMap)[])(
    'HTML event "%s" should fire a corresponding canvas event',
    (type) => {
      canvas.setViewportTransform(genericVpt);
      // select target and mock some essentials for events to fire
      const dragTarget = new IText('Drag Target', {
        originX: 'center',
        originY: 'center',
      });
      jest.spyOn(dragTarget, 'onDragStart').mockReturnValue(true);
      jest.spyOn(dragTarget, 'renderDragSourceEffect').mockImplementation();
      jest.spyOn(dragTarget, 'toJSON').mockReturnValue('Drag Target');
      canvas.add(dragTarget);
      canvas.setActiveObject(dragTarget);
      spy.mockReset();
      canvas.getSelectionElement().dispatchEvent(
        new MouseEvent('dragstart', {
          clientX: 50,
          clientY: 50,
        })
      );
      canvas.getSelectionElement().dispatchEvent(
        new MouseEvent(type, {
          clientX: 50,
          clientY: 50,
        })
      );
      expect(spy.mock.calls).toMatchSnapshot(snapshotOptions);
    }
  );

  test('getScenePoint', () => {
    const canvas = new Canvas(undefined, {
      enableRetinaScaling: true,
      width: 200,
      height: 200,
    });
    jest.spyOn(canvas, 'getRetinaScaling').mockReturnValue(200);
    const spy = jest.spyOn(canvas, 'getPointer');
    jest.spyOn(canvas.upperCanvasEl, 'getBoundingClientRect').mockReturnValue({
      width: 500,
      height: 500,
    });
    jest.spyOn(canvas.upperCanvasEl, 'width', 'get').mockReturnValue(200);
    jest.spyOn(canvas.upperCanvasEl, 'height', 'get').mockReturnValue(200);
    const ev = new MouseEvent('mousemove', {
      clientX: 50,
      clientY: 50,
    });
    const point = canvas.getScenePoint(ev);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenNthCalledWith(1, ev);
    canvas._cacheTransformEventData(ev);
    expect(point).toEqual(canvas['_absolutePointer']);
    expect(spy).toHaveBeenCalledTimes(2);
    expect(spy).toHaveBeenNthCalledWith(2, ev, true);
  });
});

it('A selected subtarget should not fire an event twice', () => {
  const target = new FabricObject();
  const group = new Group([target], {
    subTargetCheck: true,
    interactive: true,
  });
  const canvas = new Canvas();
  canvas.add(group);
  const targetSpy = jest.fn();
  target.on('mousedown', targetSpy);
  jest.spyOn(canvas, '_checkTarget').mockReturnValue(true);
  canvas.__onMouseDown({
    target: canvas.getSelectionElement(),
    clientX: 0,
    clientY: 0,
  } as unknown as TPointerEvent);
  expect(targetSpy).toHaveBeenCalledTimes(1);
});

it('should fire mouse over/out events on target', () => {
  const target = new FabricObject({ width: 10, height: 10 });
  const canvas = new Canvas();
  canvas.add(target);

  jest.spyOn(target, 'toJSON').mockReturnValue('target');

  const targetSpy = jest.spyOn(target, 'fire');
  const canvasSpy = jest.spyOn(canvas, 'fire');
  const enter = new MouseEvent('mousemove', { clientX: 5, clientY: 5 });
  const exit = new MouseEvent('mousemove', { clientX: 20, clientY: 20 });
  canvas._onMouseMove(enter);
  canvas._onMouseMove(exit);
  expect(targetSpy.mock.calls).toMatchSnapshot();
  expect(canvasSpy.mock.calls).toMatchSnapshot();
});
