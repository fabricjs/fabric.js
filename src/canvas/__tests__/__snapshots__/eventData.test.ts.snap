// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas event data HTML event "contextmenu" should fire a corresponding canvas event 1`] = `
[
  [
    "contextmenu:before",
    {
      "e": MouseEvent {
        "isTrusted": false,
      },
      "subTargets": [],
      "target": undefined,
    },
  ],
  [
    "contextmenu",
    {
      "e": MouseEvent {
        "isTrusted": false,
      },
      "subTargets": [],
      "target": undefined,
    },
  ],
]
`;

exports[`Canvas event data HTML event "dblclick" should fire a corresponding canvas event 1`] = `
[
  [
    "mouse:dblclick",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
]
`;

exports[`Canvas event data HTML event "drag" should fire a corresponding canvas event 1`] = `
[
  [
    "dragstart",
    {
      "e": MouseEvent {
        "isTrusted": false,
      },
      "target": "Drag Target",
    },
  ],
  [
    "drag",
    {
      "dragSource": "Drag Target",
      "dropTarget": undefined,
      "e": MouseEvent {
        "isTrusted": false,
      },
      "target": "Drag Target",
    },
  ],
]
`;

exports[`Canvas event data HTML event "dragend" should fire a corresponding canvas event 1`] = `
[
  [
    "dragstart",
    {
      "e": MouseEvent {
        "isTrusted": false,
      },
      "target": "Drag Target",
    },
  ],
  [
    "dragend",
    {
      "didDrop": false,
      "dragSource": "Drag Target",
      "dropTarget": undefined,
      "e": MouseEvent {
        "isTrusted": false,
      },
      "subTargets": [],
      "target": "Drag Target",
    },
  ],
  [
    "mouse:up:before",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "currentSubTargets": [],
      "currentTarget": "Drag Target",
      "e": MouseEvent {
        "isTrusted": false,
      },
      "isClick": false,
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": "Drag Target",
      "transform": null,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
  [
    "mouse:up",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "currentSubTargets": [],
      "currentTarget": "Drag Target",
      "e": MouseEvent {
        "isTrusted": false,
      },
      "isClick": false,
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": "Drag Target",
      "transform": null,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
  [
    "after:render",
    {
      "ctx": CanvasRenderingContext2D {
        "canvas": <canvas
          class="upper-canvas"
          data-fabric="top"
          draggable="true"
          height="150"
          style="position: absolute; left: 0px; top: 0px; user-select: none; width: 300px; height: 150px; cursor: move;"
          width="300"
        />,
        "createPattern": [Function],
        "drawImage": [Function],
      },
    },
  ],
]
`;

exports[`Canvas event data HTML event "dragenter" should fire a corresponding canvas event 1`] = `
[
  [
    "dragstart",
    {
      "e": MouseEvent {
        "isTrusted": false,
      },
      "target": "Drag Target",
    },
  ],
  [
    "dragenter",
    {
      "dragSource": "Drag Target",
      "e": MouseEvent {
        "isTrusted": false,
      },
      "subTargets": [],
      "target": "Drag Target",
    },
  ],
  [
    "drag:enter",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "dragSource": "Drag Target",
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "previousTarget": undefined,
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": "Drag Target",
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
]
`;

exports[`Canvas event data HTML event "dragleave" should fire a corresponding canvas event 1`] = `
[
  [
    "dragstart",
    {
      "e": MouseEvent {
        "isTrusted": false,
      },
      "target": "Drag Target",
    },
  ],
  [
    "dragleave",
    {
      "dragSource": "Drag Target",
      "e": MouseEvent {
        "isTrusted": false,
      },
      "subTargets": [],
      "target": undefined,
    },
  ],
]
`;

exports[`Canvas event data HTML event "dragover" should fire a corresponding canvas event 1`] = `
[
  [
    "dragstart",
    {
      "e": MouseEvent {
        "isTrusted": false,
      },
      "target": "Drag Target",
    },
  ],
  [
    "dragover",
    {
      "canDrop": true,
      "dragSource": "Drag Target",
      "dropTarget": "Drag Target",
      "e": MouseEvent {
        "isTrusted": false,
      },
      "subTargets": [],
      "target": "Drag Target",
    },
  ],
  [
    "drag:enter",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "canDrop": false,
      "dragSource": "Drag Target",
      "dropTarget": undefined,
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "previousTarget": undefined,
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": "Drag Target",
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
]
`;

exports[`Canvas event data HTML event "drop" should fire a corresponding canvas event 1`] = `
[
  [
    "dragstart",
    {
      "e": MouseEvent {
        "isTrusted": false,
      },
      "target": "Drag Target",
    },
  ],
  [
    "drop:before",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "didDrop": false,
      "dragSource": "Drag Target",
      "dropTarget": undefined,
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": "Drag Target",
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
  [
    "drop",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "didDrop": false,
      "dragSource": "Drag Target",
      "dropTarget": undefined,
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": "Drag Target",
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
  [
    "drop:after",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "didDrop": false,
      "dragSource": "Drag Target",
      "dropTarget": undefined,
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": "Drag Target",
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
]
`;

exports[`Canvas event data HTML event "mousedown" should fire a corresponding canvas event 1`] = `
[
  [
    "mouse:down:before",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
  [
    "mouse:down",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
]
`;

exports[`Canvas event data HTML event "mouseenter" should fire a corresponding canvas event 1`] = `
[
  [
    "mouse:over",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
]
`;

exports[`Canvas event data HTML event "mouseleave" should fire a corresponding canvas event 1`] = `[]`;

exports[`Canvas event data HTML event "mousemove" should fire a corresponding canvas event 1`] = `
[
  [
    "mouse:move:before",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
  [
    "mouse:move",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
]
`;

exports[`Canvas event data HTML event "mouseout" should fire a corresponding canvas event 1`] = `
[
  [
    "mouse:out",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "target": undefined,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
]
`;

exports[`Canvas event data HTML event "mouseover" should fire a corresponding canvas event 1`] = `[]`;

exports[`Canvas event data HTML event "mouseup" should fire a corresponding canvas event 1`] = `
[
  [
    "mouse:up:before",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "currentSubTargets": [],
      "currentTarget": undefined,
      "e": MouseEvent {
        "isTrusted": false,
      },
      "isClick": true,
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
  [
    "mouse:up",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "currentSubTargets": [],
      "currentTarget": undefined,
      "e": MouseEvent {
        "isTrusted": false,
      },
      "isClick": true,
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
]
`;

exports[`Canvas event data HTML event "wheel" should fire a corresponding canvas event 1`] = `
[
  [
    "mouse:wheel",
    {
      "absolutePointer": Point {
        "x": -30,
        "y": -13,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 50,
        "y": 50,
      },
      "scenePoint": Point {
        "x": -30,
        "y": -13,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 50,
        "y": 50,
      },
    },
  ],
]
`;

exports[`should fire mouse over/out events on target 1`] = `
[
  [
    "mouseover",
    {
      "absolutePointer": Point {
        "x": 5,
        "y": 5,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 5,
        "y": 5,
      },
      "previousTarget": undefined,
      "scenePoint": Point {
        "x": 5,
        "y": 5,
      },
      "target": "target",
      "viewportPoint": Point {
        "x": 5,
        "y": 5,
      },
    },
  ],
  [
    "mousemove",
    {
      "absolutePointer": Point {
        "x": 5,
        "y": 5,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 5,
        "y": 5,
      },
      "scenePoint": Point {
        "x": 5,
        "y": 5,
      },
      "subTargets": [],
      "target": "target",
      "transform": null,
      "viewportPoint": Point {
        "x": 5,
        "y": 5,
      },
    },
  ],
  [
    "mouseout",
    {
      "absolutePointer": Point {
        "x": 20,
        "y": 20,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "nextTarget": undefined,
      "pointer": Point {
        "x": 20,
        "y": 20,
      },
      "scenePoint": Point {
        "x": 20,
        "y": 20,
      },
      "target": "target",
      "viewportPoint": Point {
        "x": 20,
        "y": 20,
      },
    },
  ],
]
`;

exports[`should fire mouse over/out events on target 2`] = `
[
  [
    "mouse:move:before",
    {
      "absolutePointer": Point {
        "x": 5,
        "y": 5,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 5,
        "y": 5,
      },
      "scenePoint": Point {
        "x": 5,
        "y": 5,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 5,
        "y": 5,
      },
    },
  ],
  [
    "mouse:over",
    {
      "absolutePointer": Point {
        "x": 5,
        "y": 5,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 5,
        "y": 5,
      },
      "previousTarget": undefined,
      "scenePoint": Point {
        "x": 5,
        "y": 5,
      },
      "target": "target",
      "viewportPoint": Point {
        "x": 5,
        "y": 5,
      },
    },
  ],
  [
    "mouse:move",
    {
      "absolutePointer": Point {
        "x": 5,
        "y": 5,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 5,
        "y": 5,
      },
      "scenePoint": Point {
        "x": 5,
        "y": 5,
      },
      "subTargets": [],
      "target": "target",
      "transform": null,
      "viewportPoint": Point {
        "x": 5,
        "y": 5,
      },
    },
  ],
  [
    "mouse:move:before",
    {
      "absolutePointer": Point {
        "x": 20,
        "y": 20,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 20,
        "y": 20,
      },
      "scenePoint": Point {
        "x": 20,
        "y": 20,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 20,
        "y": 20,
      },
    },
  ],
  [
    "mouse:out",
    {
      "absolutePointer": Point {
        "x": 20,
        "y": 20,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "nextTarget": undefined,
      "pointer": Point {
        "x": 20,
        "y": 20,
      },
      "scenePoint": Point {
        "x": 20,
        "y": 20,
      },
      "target": "target",
      "viewportPoint": Point {
        "x": 20,
        "y": 20,
      },
    },
  ],
  [
    "mouse:move",
    {
      "absolutePointer": Point {
        "x": 20,
        "y": 20,
      },
      "e": MouseEvent {
        "isTrusted": false,
      },
      "pointer": Point {
        "x": 20,
        "y": 20,
      },
      "scenePoint": Point {
        "x": 20,
        "y": 20,
      },
      "subTargets": [],
      "target": undefined,
      "transform": null,
      "viewportPoint": Point {
        "x": 20,
        "y": 20,
      },
    },
  ],
]
`;
